src/main/kotlin/com/openlattice/graph/partioning/RepartitioningJob.kt

```diff
@@ -0,0 +1,271 @@
+                bind(countQuery, partition)
+                val rs = countQuery.executeQuery()
+                rs.next()
+                ResultSetAdapters.count(rs)
+            }
+        } catch (ex: Exception) {
+            abort { "Job $id terminated: ${ex.message}" }
+            throw ex
+        } finally {
+            publishJobState()
+        }
+    }
+
+    private fun getNeedsMigrationCount(): Long = 2*state.oldPartitions.fold(0L) { count, partition ->
+        count + getCount(idsNeedingMigrationCountSql, partition) +
+                getCount(dataNeedingMigrationCountSql, partition) +
+                getCount(edgesNeedingMigrationCountSql, partition)
+    }
+
+
+    private fun bind(ps: PreparedStatement, partition: Int = currentlyMigratingPartition) {
+        ps.setObject(1, state.entitySetId)
+        ps.setInt(2, partition)
+    }
+}
+
+
+private val REPARTITION_SELECTOR = "partitions[ 1 + ((array_length(partitions,1) + (('x'||right(${SRC_ENTITY_KEY_ID.name}::text,8))::bit(32)::int % array_length(partitions,1))) % array_length(partitions,1))]"
+
+private fun buildRepartitionColumns(ptd: PostgresTableDefinition): String {
+    return ptd.columns.joinToString(",") { if (it == PARTITION) REPARTITION_SELECTOR else it.name }
+}
+
+private val idsNeedingMigrationCountSql = """
+    SELECT count(*) 
+    FROM ${IDS.name} INNER JOIN (select id as ${ENTITY_SET_ID.name},${PARTITIONS.name} FROM ${ENTITY_SETS.name}) as es
+    WHERE ${ENTITY_SET_ID.name} = ? AND ${PARTITION.name} = ? AND ${PARTITION.name}!=$REPARTITION_SELECTOR
+""".trimIndent()
+
+private val dataNeedingMigrationCountSql = """
+    SELECT count(*) 
+    FROM ${DATA.name} INNER JOIN (select ${ID.name} as ${ENTITY_SET_ID.name}, ${PARTITIONS.name} from ${ENTITY_SETS.name}) as es 
+    USING ( ${ENTITY_SET_ID.name} )
+    WHERE ${ENTITY_SET_ID.name} = ? AND ${PARTITION.name} = ? AND ${PARTITION.name}!=$REPARTITION_SELECTOR
+""".trimIndent()
+
+private val edgesNeedingMigrationCountSql = """
+    SELECT count(*)
+    FROM ${E.name} INNER JOIN (select id as ${SRC_ENTITY_SET_ID.name}, ${PARTITIONS.name} from ${ENTITY_SETS.name}) as es
+    USING (${SRC_ENTITY_SET_ID.name})
+    WHERE ${ENTITY_SET_ID.name} = ? AND ${PARTITION.name} = ? AND ${PARTITION.name}!=$REPARTITION_SELECTOR
+""".trimIndent()
+
+private fun latestSql(
+        column: PostgresColumnDefinition,
+        comparison: PostgresColumnDefinition = column,
+        whenExcludedGreater: PostgresColumnDefinition = column,
+        otherwise: PostgresColumnDefinition = column
+): String = "${column.name} = CASE WHEN EXCLUDED.${comparison.name} > ${comparison.name} THEN EXCLUDED.${whenExcludedGreater.name} ELSE ${otherwise.name} END"
+
+
+private val REPARTITION_DATA_COLUMNS = buildRepartitionColumns(DATA)
+private val REPARTITION_IDS_COLUMNS = buildRepartitionColumns(E)
+private val REPARTITION_EDGES_COLUMNS = buildRepartitionColumns(IDS)
+
+/**
```

Comment:
Won't this take an eternity without chunking?

Reply:
Citus 9.3 has special re-partitioning logic where the nodes talk to each other. I want to try and take advantage of that first, before we start doing manual batches.

Reply:
So is the idea to test this out on a smaller/medium sized entity set before NCRIC? I'm guessing 9.3 avoids the coordinator memory limit right?

Reply:
The coordinator is not involved in re-partitioning queries if LIMIT, ORDER BY, or GROUP BY clauses are omitted. Yes, the goal would be to try this out on smaller entity sets.

Reply:
This is probably a good initial set to test repartitioning on! Most of them (save two) are suuuper small https://jira.openlattice.com/browse/LATTICE-2412?focusedCommentId=23853&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-23853